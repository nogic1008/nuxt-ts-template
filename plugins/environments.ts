import { Plugin } from '@nuxt/types'
import { config } from 'dotenv'

config()

export type EnvironmentVariables = {
  NODE_ENV: string
  /** The base URL of the app.
   * @description if the entire single page application is served under /app/,
   * then base should use the value '/app/'
   * @default '/'
   * @see https://nuxtjs.org/api/configuration-router/#base
   */
  BASE_PATH: string
}

/* eslint-disable no-process-env */
export const environments: EnvironmentVariables & {
  validate: () =>
    | { valid: true }
    | { valid: false; keys: Extract<keyof EnvironmentVariables, string>[] }
} = {
  NODE_ENV: process.env.NODE_ENV!,
  BASE_PATH: process.env.BASE_PATH || '/',
  /** Validate environments values. */
  validate() {
    const invalidKeys: string[] = Object.keys(this).filter((key) => {
      if (key === 'validate') return false
      const value: unknown = (this as any)[key]
      return value === undefined || value === null
    })
    return invalidKeys.length === 0
      ? { valid: true }
      : {
          valid: false,
          keys: invalidKeys as Extract<keyof EnvironmentVariables, string>[]
        }
  }
}
/* eslint-enable no-process-env */

declare module 'vue/types/vue' {
  interface Vue {
    $environments: EnvironmentVariables
  }
}

const environmentsPlugin: Plugin = (_, inject) => {
  const env = { ...environments }
  delete env.validate
  inject('environments', env)
}

export default environmentsPlugin
