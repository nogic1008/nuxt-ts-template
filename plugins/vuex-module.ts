import { Plugin } from '@nuxt/types'
import { VuexTypeModule, vxm } from '~/store'

declare module 'vue/types/vue' {
  interface Vue {
    $vxm: VuexTypeModule
  }
}

const vuexModulePlugin: Plugin = (_, inject) => {
  inject('vxm', vxm)
}

export default vuexModulePlugin
