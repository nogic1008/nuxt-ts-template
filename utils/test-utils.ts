/** Generate [0-9a-z]{length} string. */
export const generateRandomString = (length: number) =>
  [...Array(Number.isInteger(length) ? length : 12)]
    .map(() => (~~(Math.random() * 36)).toString(36))
    .join('')
