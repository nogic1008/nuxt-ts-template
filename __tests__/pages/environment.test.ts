import { createLocalVue, mount, shallowMount, Wrapper } from '@vue/test-utils'
import VueI18n from 'vue-i18n'

import Environment from '~/pages/environment.vue'
import { EnvironmentVariables } from '~/plugins/environments'

const localVue = createLocalVue()
localVue.use(VueI18n)

describe('pages/environment.vue', () => {
  let wrapper: Wrapper<Vue>
  const $environments: EnvironmentVariables = {
    BASE_PATH: 'foo'
  }

  beforeEach(async () => {
    wrapper = shallowMount(Environment, {
      localVue,
      mocks: { $environments }
    })
    wrapper.vm.$i18n.locale = 'en'
    await localVue.nextTick()
  })

  test('is a Vue instance', () => {
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  describe('snapshot', () => {
    test.each(['en', 'ja'])(
      'renders correctly if locale is "%s"',
      async (locale) => {
        const wrapper = mount(Environment, {
          localVue,
          mocks: { $environments }
        })
        wrapper.vm.$i18n.locale = locale
        await localVue.nextTick()
        expect(wrapper.element).toMatchSnapshot()
      }
    )
  })
})
