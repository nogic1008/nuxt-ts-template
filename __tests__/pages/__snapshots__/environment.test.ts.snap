// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/environment.vue snapshot renders correctly if locale is "en" 1`] = `
<section
  class="section"
>
  <h2
    class="title is-3 has-text-grey"
  >
    
    Client-side Environments
  
  </h2>
   
  <ul>
    <li>
      <b>
        BASE_PATH
      </b>
      
      : foo
    
    </li>
  </ul>
</section>
`;

exports[`pages/environment.vue snapshot renders correctly if locale is "ja" 1`] = `
<section
  class="section"
>
  <h2
    class="title is-3 has-text-grey"
  >
    
    クライアントサイド 環境変数
  
  </h2>
   
  <ul>
    <li>
      <b>
        BASE_PATH
      </b>
      
      : foo
    
    </li>
  </ul>
</section>
`;
